consensus/src/state_computer.rs

```diff
@@ -11,30 +12,57 @@ use consensus_types::{block::Block, executed_block::ExecutedBlock};
 use diem_crypto::HashValue;
 use diem_logger::prelude::*;
 use diem_metrics::monitor;
-use diem_types::ledger_info::LedgerInfoWithSignatures;
+use diem_types::{
+    contract_event::ContractEvent, ledger_info::LedgerInfoWithSignatures, transaction::Transaction,
+};
 use execution_correctness::ExecutionCorrectness;
 use executor_types::{Error as ExecutionError, StateComputeResult};
 use fail::fail_point;
+use futures::{SinkExt, StreamExt};
 use std::{boxed::Box, sync::Arc};
 
+type NotificationType = (
+    Box<dyn FnOnce() + Send + Sync>,
+    Vec<Transaction>,
+    Vec<ContractEvent>,
+);
+
 /// Basic communication with the Execution module;
 /// implements StateComputer traits.
 pub struct ExecutionProxy {
     execution_correctness_client: Box<dyn ExecutionCorrectness + Send + Sync>,
     mempool_notifier: Arc<dyn TxnManager>,
-    state_sync_notifier: Box<dyn ConsensusNotificationSender>,
+    state_sync_notifier: Arc<dyn ConsensusNotificationSender>,
+    async_state_sync_notifier: channel::Sender<NotificationType>,
```

Comment:
nit: it would be nice if we could move this async logic into the `ConsensusNotifier` itself, but this seems like a workable solution for now.